sdk/docs/core/az_event_pipeline.md

```diff
@@ -0,0 +1,150 @@
+
+Posted events may be either implicitly (through chains of blocking mutexes) or explicitly queued. The application or network stack must be aware of this and ensure proper event lifetime. An example could be the `AZ_EVENT_MQTT5_CONNECTION_OPEN_REQ` event posted by `az_mqtt5_connection_open_begin()` API.
+
+### Interval Timers
+
+`az_platform.h` interval timers are created and associated with the pipeline using the `_az_event_pipeline_timer` API. The timers can then be configured using non-pipeline specific `az_platform` APIs (see dependencies below).
+When the timer expires, an outbound AZ_HFSM_EVENT_TIMEOUT is posted. The event_data is a pointer to the `_az_event_pipeline_timer`. This enables pipelines to own more than one timer.
+Unless the `AZ_HFSM_EVENT_TIMEOUT` event is consumed in the current policy, it must be passed through (using send) towards other outbound policies that need to receive the event.
+
+### Pipelines enforce Run-To-Completion
+
+Data flows either inbound or outbound through the pipeline at any time. For each posted event, the data is processed on the same stack. Not all components may get invoked if the event is filtered by one of the policies.
+Each inbound or outbound operation is guarded by a mutex. There is a single mutex for the entire pipeline (i.e., an inbound and outbound operation cannot happen at the same time).
+
+### Asynchronous Mode
+
+In asynchronous mode, it is expected that all outbound and inbound events are processed immediately after being received by either the application or by the network stack. For example, in this mode of operation, it is expected that the network stack uses some form of call-backs to notify the client. The call-back will trigger a pipeline post of an inbound event.
+Because of the Run-To-Completion guarantee, all operations within a pipeline, performed on resources owned by the pipeline, are thread-safe. At the same time, because a mutex is held, the operations must not block: all operations are either CPU-bound or non-blocking I/O.
+In most cases, the CPU-bound pipeline code can run on any thread, task, or ISR.
+
+### Synchronous Mode
+
+In synchronous mode, either the outbound (e.g., API surface) or the inbound components (e.g., network stack) require a way to run their respective message loops. The Pipeline system can be configured at compile time to add synchronous message loop support through the az_hfsm_pipeline_sync_process_loop API.
+The HFSM policies must be written to support the sync loop event, AZ_HFSM_PIPELINE_EVENT_PROCESS_LOOP, which is posted first as an outbound event (to queue all network writes) then as an inbound event (to process all buffered reads).
```

Comment:
We removed the process loop API, are we going to add it again?

Reply:
I think it's useful to have this in the document - I'll add a note stating it's not implemented at this time.